ui/org.eclipse.pde.ui.tests/src/org/eclipse/ui/tests/smartimport/ProjectSmartImportTest.java

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2023 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributor:
+ *     Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package org.eclipse.ui.tests.smartimport;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.assertj.core.api.Condition;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IWorkspaceRoot;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.pde.ui.tests.util.ProjectUtils;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class ProjectSmartImportTest {
+
+	@Parameters(name = "{0}")
+	public static Object[][] projects() {
+		return new Object[][] { //
+			{ "JavaEclipseProject", List.of("org.eclipse.jdt.core.javanature") }, //
+			{ "FeatureProject", List.of("org.eclipse.pde.FeatureNature") }, //
+			{ "PlainEclipseProject", List.of() }, //
+			{ "PlainJavaProject", List.of("org.eclipse.jdt.core.javanature") }, //
+		};
+	}
+
+	@Parameter(0)
+	public String projectName;
+
+	@Parameter(1)
+	public List<String> expectedNatures;
+
+	@ClassRule
+	public static TemporaryFolder workingDirectory = new TemporaryFolder();
+
+	@BeforeClass
+	public static void setupClass() throws IOException, CoreException {
+		// Copy imported projects to temp-directory to not pollute this project
+		Path source = Path.of("tests/smartImport").toRealPath();
+		Path target = workingDirectory.getRoot().toPath();
+		try (Stream<Path> files = Files.walk(source).filter(Files::isRegularFile)) {
+			for (Path file : (Iterable<Path>) files::iterator) {
+				Path targetFile = target.resolve(source.relativize(file));
+				Files.createDirectories(targetFile.getParent());
+				Files.copy(file, targetFile);
+			}
+		}
```

Comment:
This will not work in I-builds since the test-plugin is a jar there.
Will need to generalize this to extract bundle entries, like it is done in other places.